src/QsCompiler/CompilerOptimizations/VariableRenaming.fs

```diff
@@ -0,0 +1,124 @@
+    let generateUniqueName varName =
+        let baseName = demangle varName
+        let mutable num, newName = 0, baseName
+        while numUses.ContainsKey newName do
+            num <- num + 1
+            newName <- sprintf "__qsVar%d__%s__" num baseName
+        numUses <- Map.add newName 0 numUses
+        constants <- defineVar (fun _ -> true) constants (varName, newName)
+        newName
+
+    /// Processes the initial argument tuple from the function declaration
+    let rec processArgTuple args =
+        match args with
+        | QsTupleItem item ->
+            match item.VariableName with
+            | ValidName name -> generateUniqueName name.Value |> ignore
+            | InvalidName -> ()
+        | QsTuple items -> Seq.iter processArgTuple items
+
+    do processArgTuple argTuple
+
+    /// Returns the number of times a variable is referenced
+    member internal this.getNumUses name =
+        Map.tryFind name numUses
+
+
+    override scope.Transform x =
+        if skipScope then
+            skipScope <- false
+            base.Transform x
+        else
+            constants <- enterScope constants
+            let result = base.Transform x
+            constants <- exitScope constants
+            result
+
+    override scope.Expression = { new ExpressionTransformation() with
+        override expr.Kind = { new ExpressionKindTransformation() with
+            override exprKind.ExpressionTransformation ex = expr.Transform ex
+            override exprKind.TypeTransformation t = expr.Type.Transform t
+
+            override exprKind.onIdentifier (sym, tArgs) =
+                maybe {
+                    let! name =
+                        match sym with
+                        | LocalVariable name -> Some name.Value
+                        | _ -> None
+                    let! newName = tryGetVar constants name
+                    numUses <- Map.add newName (Map.find newName numUses + 1) numUses
+                    return Identifier (LocalVariable (NonNullable<_>.New newName), tArgs)
+                } |? Identifier (sym, tArgs)
+        }
+    }
+
+    override scope.StatementKind = { new StatementKindTransformation() with
+        override stmtKind.ExpressionTransformation x = scope.Expression.Transform x
+        override stmtKind.LocationTransformation x = scope.onLocation x
+        override stmtKind.ScopeTransformation x = scope.Transform x
+        override stmtKind.TypeTransformation x = scope.Expression.Type.Transform x
+
+        override this.onSymbolTuple syms =
+            match syms with
+            | VariableName item -> VariableName (NonNullable<_>.New (generateUniqueName item.Value))
+            | VariableNameTuple items -> Seq.map this.onSymbolTuple items |> ImmutableArray.CreateRange |> VariableNameTuple
+            | InvalidItem | DiscardedItem -> syms
+
+        override this.onVariableDeclaration stm =
```

Comment:
What is the reason for this override? The only thing you changed compared to the default implementation is the order in which the lhs and rhs are processed - which should be irrelevant here.